src/classes/CW_AudienceGroupTriggerHandlerTest.cls

```diff
@@ -0,0 +1,236 @@
+		listNormalGrps.Add(listAllGrps.get(4));
+		listNormalGrps.Add(listAllGrps.get(5));
+		
+		insert listControlGrps;
+		insert listControlGrps2;
+		
+
+
+		for(integer i=0;i<2;i++){
+			listNormalGrps.get(i).Is_Control_Group__c=false;
+			listNormalGrps.get(i).Control_Group__c=listControlGrps.get(i).id;
+		}
+		insert listNormalGrps;
+	   //Test if we can chage witout ref
+	   for(CW_Audience_Group__c Rec :listControlGrps2){
+	   	Rec.Is_Control_Group__c=false;
+	   }
+	   update listControlGrps2;
+	   for(CW_Audience_Group__c Rec :listControlGrps2){
+	   	Rec.Is_Control_Group__c=true;
+	   }
+		update listControlGrps2;
+	   
+	   try{
+	   	insert listAllGrps;
+	   	System.assert(false,'Validation of Unique Name on Audience control group not rise error');
+	   }
+	   catch(Exception E){}	 
+	   
+	   
+	}
+	
+	static testMethod void controlGroupTest()
+	{
+		CW_Campaign__c camp = TestDataUtil.createCampaign();
+      	insert camp;
+        
+		CW_Audience_Group__c controlGroup = TestDataUtil.createAudienceGroup(camp.Id);
+		controlGroup.Name = 'Control1';
+		controlGroup.Is_Control_Group__c = true;
+		insert controlGroup;
+		CW_Audience_Group__c targetGroup = TestDataUtil.createAudienceGroup(camp.Id);
+		targetGroup.Name = 'Target1';
+		targetGroup.Is_Control_Group__c = false;
+		targetGroup.Control_Group__c = controlGroup.Id;
+		insert targetGroup;
+		
+		Test.startTest();
+		targetGroup.Is_Control_Group__c = true;
+		update targetGroup;
+		Test.stopTest();
+		
+		CW_Audience_Group__c reselectTarget = [select Is_Control_Group__c, Control_Group__c 
+												from CW_Audience_Group__c where id = :targetGroup.Id];
+		System.assert(reselectTarget.Is_Control_Group__c);	// group is now conrol group
+		System.assertEquals(reselectTarget.Control_Group__c,null);	// no longer has control group
+	}
+	
+	static testMethod void testCreateAGS()
+	{
+		List<CW_Schedule__c> schs = setupSchedules();	// insert later
+		
+		Test.startTest();
+		
+		// create groups with no schedules in order to avoid trigger creation
+		CW_Audience_Group__c agc = TestDataUtil.createAudienceGroup(schs.get(0).Marketing_Campaign__c);
+		agc.Name = 'test control';
+		agc.Is_Control_Group__c = true;	// control group
+		agc.Unica_Id__c = '111222';
+		insert agc;
+		
+		CW_Audience_Group__c agt = TestDataUtil.createAudienceGroup(schs.get(0).Marketing_Campaign__c);
+		agt.Name = 'test target';
+		agt.Is_Control_Group__c = false;	// control group
+		agt.Control_Group__c = agc.id;
+		agt.Unica_Id__c = '222333';
+		insert agt;
+		
+		insert schs;	// create schedules
+		
+		Test.stopTest();
+		
+		List<CW_Audience_Group_Schedule__c> agsC = 
+			[select id, Schedule__c, is_Control_Group__c from CW_Audience_Group_Schedule__c 
+				where Audience_Group__c = :agc.id order by Schedule__r.Start_Date__c asc];
+		List<CW_Audience_Group_Schedule__c> agsT = 
+			[select id, Schedule__c, is_Control_Group__c, Control_Group_Schedule__c from CW_Audience_Group_Schedule__c 
```

Comment:
Problem: Unbound SOQL query. Reason: a SOQL query can return up to 50,000 records. If this number is exceeded a runtime exception is thrown. Suggestion: cap the max number of records returned adding a LIMIT clause to the query.